File: android/src/test/java/com/amplitude/skylab/VariantTest.java
package com.amplitude.skylab;

import org.json.JSONException;
import org.json.JSONObject;
import org.junit.Assert;
import org.junit.Test;

public class VariantTest {

    @Test
    public void fromJsonObject() throws JSONException {
        {
            JSONObject jsonObject = new JSONObject();
            Variant variant = Variant.fromJsonObject(jsonObject);
            Assert.assertNull(variant.value);
            Assert.assertNull(variant.payload);
        }

        {
            JSONObject jsonObject = new JSONObject();
            jsonObject.put("value", "value");
            jsonObject.put("payload", "payload");
            Variant variant = Variant.fromJsonObject(jsonObject);
            Assert.assertEquals("value", variant.value);
            Assert.assertEquals("payload", variant.payload);
        }
    }

    @Test
    public void fromJsonObjectDeprecatedKeyField() throws JSONException {
        JSONObject jsonObject = new JSONObject();
        jsonObject.put("key", "value");
        Variant variant = Variant.fromJsonObject(jsonObject);
        Assert.assertEquals("value", variant.value);
        Assert.assertNull(variant.payload);
    }

    @Test
    public void toJson() throws JSONException {
        {
            Variant variant = new Variant("value");
            JSONObject jsonObject = new JSONObject();
            jsonObject.put("value", "value");
            Assert.assertEquals(jsonObject.toString(), variant.toJson());
        }

        {
            Variant variant = new Variant(null);
            Assert.assertEquals("{}", variant.toJson());
            Assert.assertNull(Variant.fromJson(variant.toJson()).value);
        }
    }

    @Test
    public void fromJson() {
        JSONObject jsonObject = new JSONObject();
        Assert.assertEquals(new Variant(null), Variant.fromJson(jsonObject.toString()));
    }
}

File: sdk/src/test/java/com/amplitude/experiment/VariantTest.kt
package com.amplitude.experiment

import com.amplitude.experiment.util.Logger
import com.amplitude.experiment.util.SystemLogger
import com.amplitude.experiment.util.toJson
import com.amplitude.experiment.util.toVariant
import org.json.JSONObject
import org.junit.Assert
import org.junit.Test

class VariantTest {
    init {
        Logger.implementation = SystemLogger(true)
    }

    @Test
    fun `empty json object to variant`() {
        val jsonObject = JSONObject()
        val variant = jsonObject.toVariant()
        Assert.assertNull(variant)
    }

    @Test
    fun `json object to variant`() {
        val jsonObject = JSONObject()
        jsonObject.put("key", "key")
        jsonObject.put("value", "value")
        jsonObject.put("payload", "payload")
        jsonObject.put("expKey", "expKey")
        jsonObject.put("metadata", emptyMap<Any, Any>())
        val variant = jsonObject.toVariant()
        Assert.assertNotNull(variant)
        Assert.assertEquals("key", variant!!.key)
        Assert.assertEquals("value", variant.value)
        Assert.assertEquals("payload", variant.payload)
        Assert.assertEquals("expKey", variant.expKey)
        Assert.assertEquals(mapOf("experimentKey" to "expKey"), variant.metadata)
    }

    @Test
    fun `variant to json object`() {
        run {
            val variant = Variant("value", null, "expKey", "key", emptyMap())
            val jsonObject = JSONObject()
            jsonObject.put("key", "key")
            jsonObject.put("value", "value")
            jsonObject.put("expKey", "expKey")
            jsonObject.put("metadata", emptyMap<Any, Any>())
            Assert.assertEquals(jsonObject.toString(), variant.toJson())
        }
    }
}
